tests/test_project/tests/test_password_history.py

```diff
@@ -100,3 +101,177 @@ def test_change_number_hasher_iterations(self):
+                order_by('pk')
+        )
+        self.assertEqual(
+            current_passwords,
+            [ph2.pk, ph3.pk],
+            msg='Only the passwords of the first user can be deleted'
+        )
+
+        current_passwords = list(
+            PasswordHistory.objects. \
+                filter(user_config__user=user2). \
+                values_list('pk', flat=True). \
+                order_by('pk')
+        )
+        self.assertEqual(
+            current_passwords,
+            [ph1_u2.pk, ph2_u2.pk, ph3_u2.pk],
+            msg='Password history of the other user must be unchanged'
+        )
+
+    def test_validation_lookup__delete_old_passwords__multiple_UserPasswordHistoryConfig(self):
+        user1 = self.create_user(1)
+        PasswordHistory.objects.all().delete()
+
+        user1_uphc1 = UserPasswordHistoryConfig.objects.filter(user=user1)[0]
+        user1_uphc2 = UserPasswordHistoryConfig(user=user1, salt='qwerty', iterations=10)
+        user1_uphc2.save()
+
+        ph1 = PasswordHistory(user_config=user1_uphc1, password='2 user1 hash1')  # to delete
+        ph1.save()
+        ph2 = PasswordHistory(user_config=user1_uphc2, password='1 user1 hash2')
+        ph2.save()
+        ph3 = PasswordHistory(user_config=user1_uphc1, password='3 user1 hash3')
+        ph3.save()
+        upv = UniquePasswordsValidator(last_passwords=2)
+        upv.delete_old_passwords(user1)
+        current_passwords = list(
+            PasswordHistory.objects.filter(user_config__user=user1).values_list('pk', flat=True).order_by('pk'))
+        self.assertEqual(
+            current_passwords,
+            [ph2.pk, ph3.pk],
+            msg='Only the oldest password can be deleted = ph1'
+        )
+        PasswordHistory.objects.all().delete()
+
+        ph1 = PasswordHistory(user_config=user1_uphc2, password='2 user1 hash1')  # to delete
+        ph1.save()
+        ph2 = PasswordHistory(user_config=user1_uphc1, password='1 user1 hash2')
+        ph2.save()
+        ph3 = PasswordHistory(user_config=user1_uphc1, password='3 user1 hash3')
+        ph3.save()
+        upv = UniquePasswordsValidator(last_passwords=2)
+        upv.delete_old_passwords(user1)
+        current_passwords = list(
+            PasswordHistory.objects.filter(user_config__user=user1).values_list('pk', flat=True).order_by('pk'))
+        self.assertEqual(
+            current_passwords,
+            [ph2.pk, ph3.pk],
+            msg='Only the oldest password can be deleted = ph1'
+        )
+        PasswordHistory.objects.all().delete()
+
+        ph1 = PasswordHistory(user_config=user1_uphc2, password='2 user1 hash1')  # to delete
+        ph1.save()
+        ph2 = PasswordHistory(user_config=user1_uphc1, password='1 user1 hash2')
+        ph2.save()
+        ph3 = PasswordHistory(user_config=user1_uphc2, password='3 user1 hash3')
+        ph3.save()
+        upv = UniquePasswordsValidator(last_passwords=2)
```

Comment:
Maybe here is a good place. This is the only test with several `UniquePasswordsValidator`, but all of them have last_passwords=2. Could you add test cases for changing from 2 to 1, and from 2 to 3, for instance?

- 2 to 1: We have passwordhistorys A and B. A is now valid and/or gets deleted as it's outside last_passwords.
- 3 to 2: We have passwordhistorys A and B. They remain invalid (there are no side effects to having a large last_passwords)
- 0 to 2: Test whether passwords are stored at all, when set to 2 should it consider the past entries?

However you consider! :)

Reply:
I've added a few more tests that probably cover everything I can think of already.